### PR TITLE
Add several algorithm routines for triangles

### PR DESCRIPTION
**Pull request by aprokop**
This PR adds several routines and updates:
- `centroid(Triangle)`
- `expand(Box, Triangle)`
- `intersects(Point, Triangle)` (2D only)
- changes all `centroid` and `intersects` calls to `constexpr`

The first two are necessary to be able to construct a hierarchy based on triangles. The third one is useful for the 2D triangle intersection example and applications (like XGC/Tokamak). The last one is a drive-by change

### CODE REVIEW

**Comment by aprokop** [comment]
Two builds have not started, but the ones that did passed.

File: src/geometry/ArborX_DetailsAlgorithms.hpp [hunk 1]
```diff
@@ -397,6 +412,45 @@ struct intersects<PointTag, SphereTag, Point, Sphere>
   }
 };
 
+template <typename Point, typename Triangle>
+struct intersects<PointTag, TriangleTag, Point, Triangle>
+{
+  KOKKOS_FUNCTION static bool apply(Point const &point,
```
Comment (dalg24) [inline src/geometry/ArborX_DetailsAlgorithms.hpp:418]: Any particular reason why this is not constexpr?
Same comment for expand.

File: test/tstDetailsAlgorithms.cpp [hunk 2]
```diff
@@ -215,6 +236,12 @@ BOOST_AUTO_TEST_CASE(centroid)
   BOOST_TEST(center[0] == -5.0);
   BOOST_TEST(center[1] == 5.0);
   BOOST_TEST(center[2] == 15.0);
+
+  Triangle triangle{{{0, 0, -2}}, {{3, 0, 1}}, {{0, 3, 1}}};
+  auto tri_center = returnCentroid(triangle);
+  BOOST_TEST(tri_center[0] == 1);
+  BOOST_TEST(tri_center[1] == 1);
+  BOOST_TEST(tri_center[2] == 0);
```
Comment (dalg24) [inline test/tstDetailsAlgorithms.cpp:244]: Add a 2D triangle test?

File: test/tstDetailsAlgorithms.cpp [hunk 3]
```diff
@@ -157,6 +159,19 @@ BOOST_AUTO_TEST_CASE(intersects)
   BOOST_TEST(intersects(Point{-1., 0., 0.}, sphere));
   BOOST_TEST(intersects(Point{-0.6, -0.8, 0.}, sphere));
   BOOST_TEST(!intersects(Point{-0.7, -0.8, 0.}, sphere));
+
+  // triangle
+  using Point2 = ArborX::ExperimentalHyperGeometry::Point<2>;
+  constexpr ArborX::ExperimentalHyperGeometry::Triangle<2> triangle{
+      {{0, 0}}, {{1, 0}}, {{0, 2}}};
+  BOOST_TEST(intersects(Point2{{0, 0}}, triangle));
+  BOOST_TEST(intersects(Point2{{1, 0}}, triangle));
```
Comment (dalg24) [inline test/tstDetailsAlgorithms.cpp:168]: Please test with the 3rd support point too.

Comment (dalg24) [reply test/tstDetailsAlgorithms.cpp:168]: What about points on the edges like (.5, 0.)?

Comment (aprokop) [reply test/tstDetailsAlgorithms.cpp:168]: > What about points on the edges like (.5, 0.)?

(0.5, 1) is on the edge.

**Comment by aprokop** [comment]
@dalg24 updated

File: test/tstDetailsAlgorithms.cpp (hunk 3 above)
Comment (dalg24) [reply test/tstDetailsAlgorithms.cpp:168]: maybe we should have center of all 3 edges?

File: src/geometry/ArborX_DetailsAlgorithms.hpp (hunk 1 above)
Comment (aprokop) [reply src/geometry/ArborX_DetailsAlgorithms.hpp]: `expand` changes the objects, with the first argument being non-const.

File: test/tstDetailsAlgorithms.cpp (hunk 3 above)
Comment (aprokop) [reply test/tstDetailsAlgorithms.cpp:168]: Sure.

File: src/geometry/ArborX_DetailsAlgorithms.hpp (hunk 1 above)
Comment (dalg24) [reply src/geometry/ArborX_DetailsAlgorithms.hpp]: OK we can fix later. This will involve more changes.

**Comment by aprokop** [comment]
Builds pass, just a warning from a container build.

**Review by dalg24 — APPROVED** [review]

**Pull request merged**